### PR TITLE
Add ProcessProcedure.

### PR DESCRIPTION
**Pull request by yageek**
The class has been largely inspired by the original `TaskOperation`.
Refers to #568.

### CODE REVIEW

File: Sources/Mac/Process.swift [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+     Initializes ProcessProcedure with a Process.
+     - parameter task: the Process
+     - parameter processDidExitCleanly: a ProcessDidExitCleanly closure with a default.
+     */
+    public init(process: Process, processDidExitCleanly: @escaping ProcessDidExitCleanly = ProcessProcedure.defaultProcessDidExitCleanly) {
+        self.process = process
+        self.processDidExitCleanly = processDidExitCleanly
+        super.init()
+
+        addWillCancelBlockObserver { [unowned self] procedure, errors in
```
Comment (danthorpe) [inline Sources/Mac/Process.swift]: So, there is no need here to capture `self` in these observer blocks anymore. In this case `procedure` is already guaranteed to be the receiver, with full type safety. So you can just do this:

```swift
addWillCancelBlockObserver { procedure, errors in
    guard procedure.process.isRunning else { return }
    procedure.process.terminate()
}
```

File: Sources/Mac/Process.swift [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    open override func execute() {
+        guard let _ = process.launchPath else {
+            finish(withError: Error.LaunchPathNotSet)
+            return
+        }
+
+        let previousTerminationHandler = process.terminationHandler
+
+        process.terminationHandler = { [unowned self] task in
```
Comment (danthorpe) [inline Sources/Mac/Process.swift]: I think here, we'd prefer this to be changed to `[weak self]`, throughout the framework we're changing all of these sort of captures to be weak, as it's not necessarily guaranteed that `self` will not have been deallocated when the closure is invoked. So, something like:

```swift
process.terminationHandler = { [weak self] task in
    guard let strongSelf = self else { return }
    // etc etc
}
```

File: Sources/Mac/Process.swift [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+import Foundation
+import ProcedureKit
+
+open class ProcessProcedure: Procedure {
+
+    /// Error type for ProcessProcedure
+    public enum Error: Swift.Error, Equatable {
+        case LaunchPathNotSet
+        case TerminationReason(Process.TerminationReason)
```
Comment (danthorpe) [inline Sources/Mac/Process.swift]: Can we change these cases to be Swift 3.0 style, which is camel case. 😀

File: Sources/Mac/Process.swift [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            else {
+                self.finish(withError: Error.TerminationReason(self.process.terminationReason))
+            }
+        }
+
+        process.launch()
+    }
+}
+
+public func == (lhs: ProcessProcedure.Error, rhs: ProcessProcedure.Error) -> Bool {
```
Comment (danthorpe) [inline Sources/Mac/Process.swift]: Not really sure what is going on here, but in Swift 3.0, conforming to equatable means putting this as a static function inside the type?

Comment (yageek) [inline Sources/Mac/Process.swift]: I was not aware about that change. I will fix that.

**Comment by yageek** [comment]
Regarding the implementation of `Equatable` protocol, it is still possible to implement it in the global scope. I tried with the `Identifiable` protocol and implement it like:

```swift
public protocol Identifiable {
    var identifier: UUID { get }
}

public func ==<T: Identifiable>(lhs: T, rhs: T) -> Bool {
    return lhs.identifier == rhs.identifier
}

```

**Comment by danthorpe** [comment]
@yageek - yeah, I guess that's a little bit confusing. 🤔 So far, I've been adding the static functions. And I think I prefer it, as it helps keep implementations collected together.

**Comment by danthorpe** [comment]
Anyway, this is great! Thanks @yageek!

**Pull request merged**